turbopack/crates/turbopack-ecmascript/src/global_module_id_strategy.rs

```diff
@@ -164,58 +163,50 @@ pub async fn children_modules_idents(
     Ok(PreprocessedChildrenIdents { modules_idents }.cell())
 }
 
+const JS_MAX_SAFE_INTEGER: u64 = (1u64 << 53) - 1;
+
 // Note(LichuAcu): This could be split into two functions: one that merges the preprocessed module
 // ids and another that generates the final, optimized module ids. Thoughts?
 pub async fn merge_preprocessed_module_ids(
     preprocessed_module_ids: Vec<Vc<PreprocessedChildrenIdents>>,
-) -> Result<HashMap<RcStr, Vc<ModuleId>>> {
-    let mut module_id_map: HashMap<RcStr, Vc<ModuleId>> = HashMap::new();
-    let mut used_ids: HashSet<u64> = HashSet::new();
+) -> Result<IndexMap<RcStr, Vc<ModuleId>>> {
+    let mut merged_module_ids = IndexMap::new();
 
     for preprocessed_module_ids in preprocessed_module_ids {
-        for (module_ident, full_hash) in preprocessed_module_ids.await?.modules_idents.iter() {
-            process_module(
-                module_ident.clone(),
-                *full_hash,
-                &mut module_id_map,
-                &mut used_ids,
-            )
-            .await?;
+        for (module_ident, full_hash) in &preprocessed_module_ids.await?.modules_idents {
+            merged_module_ids.insert(module_ident.clone(), *full_hash);
         }
     }
 
-    Ok(module_id_map)
-}
-
-pub async fn process_module(
-    ident_str: RcStr,
-    full_hash: u64,
-    id_map: &mut HashMap<RcStr, Vc<ModuleId>>,
-    used_ids: &mut HashSet<u64>,
-) -> Result<()> {
-    if id_map.contains_key(&ident_str) {
-        return Ok(());
-    }
-
-    let mut trimmed_hash = full_hash % 10;
-    let mut power = 10;
-    while used_ids.contains(&trimmed_hash) {
-        if power >= u64::MAX / 10 {
-            // We don't want to take the next power as it would overflow
-            if used_ids.contains(&full_hash) {
-                return Err(anyhow::anyhow!("This is a... 64-bit hash collision?"));
-            }
-            trimmed_hash = full_hash;
-            break;
+    // 5% fill rate, as done in Webpack
+    // https://github.com/webpack/webpack/blob/27cf3e59f5f289dfc4d76b7a1df2edbc4e651589/lib/ids/IdHelpers.js#L366-L405
+    let optimal_range = merged_module_ids.len() * 20;
+    let digit_mask = std::cmp::min(
+        10u64.pow((optimal_range as f64).log10().ceil() as u32),
+        JS_MAX_SAFE_INTEGER,
+    );
+
+    let mut module_id_map: IndexMap<RcStr, Vc<ModuleId>> = IndexMap::new();
```

Comment:
(More broadly than just this PR) I don't know all the places `Vc<ModuleId>` is used, but we should probably avoid storing collections of `ModuleId`s as a bunch of tiny `Vc`s, and just store and pass them [around by value](https://turbopack-rust-docs.vercel.sh/rustdoc/turbo_tasks/struct.Value.html#impl-TaskInput-for-Value%3CT%3E) in most cases.

A `ModuleId` is

```
#[turbo_tasks::value(shared)]
#[derive(Debug, Clone, Hash, Ord, PartialOrd, DeterministicHash)]
#[serde(untagged)]
pub enum ModuleId {
    Number(u64),
    String(RcStr),
}
```

which is 16 bytes (and could be 8 bits if `Number` was `NonZeroU64`, thanks to the niche optimization!). A `Vc` is also 16 bytes (plus all the overhead of allocating a cell in the backend for it, if it wasn't already in a cell).

Reply:
It need to be a Vc to be returned from a task. And we need to return a single Vc from `get_module_id`.

But it should be `IndexMap<RcStr, ModuleId>` here and the ModuleId **must** be celled in `get_module_id`